[kube-prometheus-stack] Use namespaceOverrides when set

sathieu opened the pull request:
#### What this PR does / why we need it:

Uses `grafana.namespaceOverride`, `kube-state-metrics.namespaceOverride` and `prometheus-node-exporter.namespaceOverride` when set, for created resource.

(includes #1342).

#### Checklist
- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

vsliouniaev commented on charts/kube-prometheus-stack/templates/grafana/dashboards-1.14/etcd.yaml:
```diff
@@ -8,7 +8,7 @@ https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-promet
 kind: ConfigMap
 metadata:
-  namespace: {{ template "kube-prometheus-stack.namespace" . }}
+  namespace: {{ template "grafana.namespace" . }}
```
These files are generated from a python script - please update this and re-run to regenerate the dashboards.

sathieu replied:
@vsliouniaev Done. Please re-trigger the CI.

vsliouniaev commented on charts/kube-prometheus-stack/templates/grafana/dashboards-1.14/alertmanager-overview.yaml:
```diff
@@ -8,8 +8,8 @@ https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-promet
-  namespace: {{ template "kube-prometheus-stack.namespace" . }}
-  name: {{ printf "%s-%s" (include "kube-prometheus-stack.fullname" $) "alertmanager-overview" | trunc 63 | trimSuffix "-" }}
+  namespace: {{ template "grafana.namespace" . }}
+  name: {{ printf "%s-%s" (include "grafana.fullname" $) "alertmanager-overview" | trunc 63 | trimSuffix "-" }}
```
Did you intend to change the name of the configmap as well here?

sathieu replied:
@vsliouniaev No, sorry. Fixed.

vsliouniaev reviewed (commented):
Sorry, noticed another potential issue with namespace selectors in serivcemonitors: 

https://github.com/prometheus-community/helm-charts/pull/1343/files#diff-ef9695f03f05e41615990a96c6a839fd0858697075287a8be909cc1acc18509aR16-R20

Could you try testing this end-to-end with all possible namespace overrides set to different values and see if everything remains fully operational?

sathieu commented:
@vsliouniaev Actually, I have a problem when doing Helm template (from ArgoCD):
```
`helm template . [...]` failed exit status 1: Error: template: prometheus-stack/charts/kube-prometheus-stack/templates/_helpers.tpl:103:16: executing "grafana.namespace" at <.Values.grafana.namespaceOverride>: nil pointer evaluating interface {}.namespaceOverride Use --debug flag to render out invalid YAML
```
The CI [has the same message](https://github.com/prometheus-community/helm-charts/pull/1343/checks?check_run_id=3620964656#step:8:53).

Also, your link is not pointing to a specific file. Could you update it?

sathieu commented:
@vsliouniaev I found the cause: [from the doc](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/):

> Recall how the labels on templates are globally shared. Thus, the labels chart can be included from any other chart.

... and defining a template twice leads to mixed scope.

To fix this, I removed the templates from `kube-prometheus-stack`, so that subchart's templates are used.

sathieu commented:
In can confirm that it works:

```yaml
$ helm template  --set grafana.namespaceOverride=grafana --set kube-state-metrics.namespaceOverride=kube-state-metrics --set prometheus-node-exporter.namespaceOverride=prometheus-node-exporter .  | grep -A 5 'kind: ServiceMonitor'
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-alertmanager
  namespace: default
  labels:
    app: kube-prometheus-stack-alertmanager
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-coredns
  namespace: default
  labels:
    app: kube-prometheus-stack-coredns
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-apiserver
  namespace: default
  labels:
    app: kube-prometheus-stack-apiserver
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kube-controller-manager
  namespace: default
  labels:
    app: kube-prometheus-stack-kube-controller-manager
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kube-etcd
  namespace: default
  labels:
    app: kube-prometheus-stack-kube-etcd
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kube-proxy
  namespace: default
  labels:
    app: kube-prometheus-stack-kube-proxy
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kube-scheduler
  namespace: default
  labels:
    app: kube-prometheus-stack-kube-scheduler
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kube-state-metrics
  namespace: kube-state-metrics
  labels:
    app: kube-prometheus-stack-kube-state-metrics
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-kubelet
  namespace: default
  labels:
    app: kube-prometheus-stack-kubelet
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-node-exporter
  namespace: prometheus-node-exporter
  labels:
    app: kube-prometheus-stack-node-exporter
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-grafana
  namespace: grafana
  labels:
    app: kube-prometheus-stack-grafana
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-operator
  namespace: default
  labels:
    app: kube-prometheus-stack-operator
--
kind: ServiceMonitor
metadata:
  name: RELEASE-NAME-kube-promethe-prometheus
  namespace: default
  labels:
    app: kube-prometheus-stack-prometheus
```

vsliouniaev commented:
There are namespace selectors in a few servicemonitors that I think will cause issues: https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/exporters/kube-state-metrics/serviceMonitor.yaml#L50-L54

sathieu commented:
thanks @vsliouniaev I've removed `kubeStateMetrics.serviceMonitor.namespaceOverride` and bumped major for this.

I've also fixed the prometheus Datasource.

Currently testing ...

sathieu commented:
At 4f92516, with `grafana.namespaceOverride`, `kube-state-metrics.namespaceOverride` and `prometheus-node-exporter.namespaceOverride`, I have tested:

- ✔️ Prometheus Datasource is set-up and working
- ✔️ Dashboards are created
- ✔️ All tested dashboards have filled graphs, including
  - ✔️ Node Exporter / Nodes (`prometheus-node-exporter.namespaceOverride`)

sathieu commented:
@vsliouniaev Please review again.

NB: I've rebased to resolve a merge conflict.

vsliouniaev commented:
From the code, I can see there is an inconsistency in the namespaces in charts/kube-prometheus-stack/templates/exporters/node-exporter/servicemonitor.yaml.

sathieu commented:
@vsliouniaev Thanks. Fixed. back to you 🏓 !

vsliouniaev commented:
This generally looks like it's working. Not sure what the use-case is for such an added complication to the chart.

sathieu commented:
@vsliouniaev I think it's a good idea to separate the components in different namespaces. Namespaces are security boundaries in k8s, and usage of network policies or roles is easier. Monitoring is also easier.

Also, I'm not the only one using namespace overrides (See older bugs #984, #820), this PR fixes all remaining problems about nsOverrides.

Note: I've resolved conflict and force-pushed.

sathieu commented:
Thanks @vsliouniaev !

james-callahan commented:
This breaks namespace independence. 
`namespaceSelector` shouldn't be set unless it's pointing at a different namespace
